generator/restriction_collector.cpp

```diff
@@ -0,0 +1,218 @@
+
+    numbers.push_back(number);
+  }
+  return true;
+}
+}  // namespace
+
+namespace routing
+{
+RestrictionCollector::RestrictionCollector(string const & restrictionPath,
+                                           string const & featureId2OsmIdsPath)
+{
+  if (!ParseFeatureId2OsmIdsMapping(featureId2OsmIdsPath))
+    return;
+
+  if (!ParseRestrictions(restrictionPath))
+  {
+    m_restrictions.clear();
+    return;
+  }
+  ComposeRestrictions();
+  RemoveInvalidRestrictions();
+  LOG(LINFO, ("m_restrictions.size() =", m_restrictions.size()));
+}
+
+bool RestrictionCollector::IsValid() const
+{
+  return !m_restrictions.empty() &&
+         find_if(begin(m_restrictions), end(m_restrictions),
+                 [](Restriction const & r) { return !r.IsValid(); }) == end(m_restrictions);
+}
+
+bool RestrictionCollector::ParseFeatureId2OsmIdsMapping(string const & featureId2OsmIdsPath)
+{
+  ifstream featureId2OsmIdsStream(featureId2OsmIdsPath);
+  if (featureId2OsmIdsStream.fail())
+    return false;
+
+  while (featureId2OsmIdsStream)
+  {
+    string line;
+    if (!getline(featureId2OsmIdsStream, line))
+      return true;
+
+    istringstream lineStream(line);
+    vector<uint64_t> ids;
+    if (!ParseLineOfNumbers(lineStream, ids))
+      return false;
+
+    if (ids.size() <= 1)
+      return false;  // Every line should contain at least feature id and osm id.
+
+    Restriction::FeatureId const featureId = static_cast<Restriction::FeatureId>(ids.front());
+    ids.erase(ids.begin());
+    AddFeatureId(featureId, ids);
+  }
+  return true;
+}
+
+bool RestrictionCollector::ParseRestrictions(string const & restrictionPath)
+{
+  ifstream restrictionsStream(restrictionPath);
+  if (restrictionsStream.fail())
+    return false;
+
+  while (restrictionsStream)
+  {
+    string line;
+    if (!getline(restrictionsStream, line))
+      return true;
+    istringstream lineStream(line);
+    string typeStr;
+    getline(lineStream, typeStr, ',');
+    Restriction::Type type;
+    if (!FromString(typeStr, type))
+      return false;
+
+    vector<uint64_t> osmIds;
+    if (!ParseLineOfNumbers(lineStream, osmIds))
+      return false;
+
+    AddRestriction(type, osmIds);
+  }
+  return true;
+}
+
+void RestrictionCollector::ComposeRestrictions()
+{
+  // Going through all osm id saved in |m_restrictionIndex| (mentioned in restrictions).
+  size_t const restrictionSz = m_restrictions.size();
+  for (pair<uint64_t, Index> const & osmIdAndIndex : m_restrictionIndex)
+  {
+    Index const & index = osmIdAndIndex.second;
+    CHECK_LESS(index.m_restrictionNumber, restrictionSz, ());
+    Restriction & restriction = m_restrictions[index.m_restrictionNumber];
+    CHECK_LESS(index.m_linkNumber, restriction.m_links.size(), ());
+
+    uint64_t const & osmId = osmIdAndIndex.first;
+    // Checking if there's an osm id belongs to a restriction is saved only once as feature id.
+    auto const rangeId = m_osmIds2FeatureId.equal_range(osmId);
```

Comment:
When you need a range of equal (in accordance with some predicate) values, map is better.